### PR TITLE
[MM-34233] Only refresh App bindings if apps plugin is enabled

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

This PR solves the issue of unnecessary requests going to the Apps plugin when it is disabled, as well as solving invalidation of bindings when a request to fetch new bindings fails. @larkox This will need to be sync'd up with your PR organizing the RHS bindings.

#### Ticket Link

Fixes https://mattermost.atlassian.net/browse/MM-34233
Fixes https://mattermost.atlassian.net/browse/MM-34710

#### Related Pull Requests

~~This PR relies on the Apps plugin having a dummy webapp plugin, to hook into the plugin enable sync system already in place for the web application. The mobile application will use the same API endpoints and websocket messages to determine if the Apps plugin is running.~~

Edit: We are instead using the Apps plugin's websocket messages to signal that the plugin has been enabled or disabled. See https://github.com/mattermost/mattermost-plugin-apps/pull/253

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: actions/websocket_actions.jsx [hunk 1]
```diff
@@ -1223,6 +1227,8 @@ function handleChannelViewedEvent(msg) {
 
 export function handlePluginEnabled(msg) {
     const manifest = msg.data.manifest;
+    dispatch({type: ActionTypes.RECEIVED_WEBAPP_PLUGIN, data: manifest});
```
Comment (larkox) [inline actions/websocket_actions.jsx:1230]: Shouldn't we do something also on plugin disabled?

File: packages/mattermost-redux/src/reducers/entities/apps.ts [hunk 2]
```diff
@@ -13,6 +13,9 @@ import {validateBindings} from 'mattermost-redux/utils/apps';
 
 export function bindings(state: AppBinding[] = [], action: GenericAction): AppBinding[] {
     switch (action.type) {
+    case AppsTypes.FAILED_TO_FETCH_APP_BINDINGS: {
+        return [];
+    }
```
[the remark this answers is not in the document]
Comment (larkox) [reply packages/mattermost-redux/src/reducers/entities/apps.ts:18]: I agree with you. We should invalidate no matter the error. I am trying to fabricate an scenario where this would be a problem, but I cannot think of any real enough. If we get to any problem in the future, the fix is simple.

File: packages/mattermost-redux/src/types/plugins.ts [hunk 3]
```diff
@@ -134,3 +139,37 @@ export type MarketplacePlugin = { // TODO remove this in favour of the definitio
     manifest: PluginManifest;
     installed_version?: string;
 }
+
+export type PluginsState = {
```
Comment (larkox) [inline packages/mattermost-redux/src/types/plugins.ts:143]: This is already defined in `/types/store/` `index.ts` and `plugins.ts` (not in redux). Not 100% sure the value of adding it here (but I like that it is more deeply typed than the available one).

File: actions/websocket_actions.jsx (hunk 1 above)
Comment (larkox) [inline actions/websocket_actions.jsx:1230]: Should we trigger a fetch bindings when the status of the apps plugin is changed to enabled?
Should we trigger a binding clear when the status of the apps plugin is changed to disabled?

**Review by larkox — COMMENTED** [review]
In general LGTM. Just a few comments to start some discussion.

File: actions/websocket_actions.jsx [hunk 4]
```diff
@@ -184,23 +184,27 @@ export function reconnect(includeWebSocket = true) {
         timestamp: Date.now(),
     });
 
-    loadPluginsIfNecessary();
+    const state = getState();
+    const currentTeamId = state.entities.teams.currentTeamId;
+
+    loadPluginsIfNecessary().then(() => {
+        if (currentTeamId) {
```
Comment (hmhealey) [inline actions/websocket_actions.jsx:191]: Since this is run asynchronously, we should call `getState` in here instead of outside. `currentTeamId` may have changed between the time the websocket reconnects to when the plugins are reloaded

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: actions/websocket_actions.jsx (hunk 1 above)
Comment (mickmister) [reply actions/websocket_actions.jsx:1230]: This is already handled as `handlePluginDisabled` calls `removePlugin`, which calls `removeWebappPlugin` which dispatches the action to change the redux store: https://github.com/mattermost/mattermost-webapp/blob/f9b4163345d80fcb6d01a574c41c6c3c227e2ec4/plugins/actions.js#L8-L13

For some reason it wasn't being handled when a plugin is enabled. I think the state of the reducer doesn't really matter to most of the application, so it was fine to be out of sync in some way, but now we need it to be correct.

File: packages/mattermost-redux/src/types/plugins.ts (hunk 3 above)
Comment (mickmister) [reply packages/mattermost-redux/src/types/plugins.ts:143]: This actually also (at least partially) needs to go to the mobile repo as well, as it will also be using this to determine if the Apps plugin is currently available. It seems weird, but we are using the existing mechanism of enabling/disabling webapp plugins at runtime for the sole purpose of keeping states in sync with the Apps plugin's state.

**Comment by mickmister** [comment]
@larkox @hmhealey I've responded to your questions, I'll get the changes implemented soon

File: actions/websocket_actions.jsx (hunk 1 above)
Comment (mickmister) [reply actions/websocket_actions.jsx:1231]: Done :+1:

Comment (mickmister) [reply actions/websocket_actions.jsx:1231]: I decided to keep this change in since it's a bug fix

**Comment by mickmister** [comment]
The logic of this PR has changed to listen for the websocket messages defined in https://github.com/mattermost/mattermost-plugin-apps/pull/253, that lets the frontend know when the Apps plugin is enabled or disabled. This was decided instead of using the webapp plugin route for several reasons.

See this thread for full context https://community-daily.mattermost.com/core/pl/wswoayjrxbbgznsj6ewxip8jie

File: actions/websocket_actions.jsx [hunk 5]
```diff
@@ -1223,6 +1228,8 @@ function handleChannelViewedEvent(msg) {
 
 export function handlePluginEnabled(msg) {
     const manifest = msg.data.manifest;
+    dispatch({type: ActionTypes.RECEIVED_WEBAPP_PLUGIN, data: manifest});
```
Comment (larkox) [inline actions/websocket_actions.jsx:1231]: Is this change still needed?

File: actions/websocket_actions.test.jsx [hunk 6]
```diff
@@ -845,18 +845,29 @@ describe('handlePluginEnabled/handlePluginDisabled', () => {
             const mockComponent = 'mockRootComponent';
             registery.registerRootComponent(mockComponent);
 
-            const dispatchArg = store.dispatch.mock.calls[0][0];
+            let dispatchArg = store.dispatch.mock.calls[0][0];
```
Comment (larkox) [inline actions/websocket_actions.test.jsx:848]: Are this test changes still needed?

Comment (larkox) [reply actions/websocket_actions.test.jsx:848]: Should we add tests for the new events?

**Review by larkox — APPROVED** [review]
LGTM, just some potential cleanup.

File: actions/websocket_actions.jsx (hunk 5 above)
Comment (mickmister) [reply actions/websocket_actions.jsx:1231]: @larkox I kept this here since it's a bug fix

File: actions/websocket_actions.test.jsx (hunk 6 above)
Comment (mickmister) [reply actions/websocket_actions.test.jsx:848]: @larkox I think the test adds value so I decided to keep it. Yes the new functionality should have tests. Thanks for suggesting this

**Review by hmhealey — APPROVED** [review]

File: packages/mattermost-redux/src/reducers/entities/apps.ts [hunk 7]
```diff
@@ -13,15 +13,50 @@ import {validateBindings} from 'mattermost-redux/utils/apps';
...
+        return true;
+    }
+    case AppsTypes.APPS_PLUGIN_DISABLED: {
+        return false;
+    }
+    case AppsTypes.RECEIVED_APP_BINDINGS: {
+        return true;
+    }
+    case AppsTypes.FAILED_TO_FETCH_APP_BINDINGS: {
+        return false;
```
Comment (mickmister) [inline packages/mattermost-redux/src/reducers/entities/apps.ts:51]: @larkox We may want to check if we received a 404 here, as there may be another reason why this failed. If the Apps plugin crashes, we will later receive a "plugin enabled" event to then know that it's available, so in that case it's better not to check.

The case we would want to avoid is there being a problem with the Apps plugin that causes it to return a non-404 error, and has the problem corrected without the plugin restarting, because then the frontend is left in a state where it thinks the plugin is unavailable, even though it is available after the problem is corrected.

File: actions/websocket_actions.jsx [hunk 8]
```diff
@@ -1412,13 +1418,30 @@ function handleCloudPaymentStatusUpdated() {
 function handleRefreshAppsBindings() {
     return (doDispatch, doGetState) => {
         const state = doGetState();
-        if (appsEnabled(state)) {
-            doDispatch(fetchAppBindings(getCurrentUserId(state), getCurrentChannelId(state)));
-        }
+        doDispatch(fetchAppBindings(getCurrentUserId(state), getCurrentChannelId(state)));
```
Comment (mickmister) [inline actions/websocket_actions.jsx:1421]: @larkox Notice that I removed the `if (appsEnabled(state))` check here. This is because we don't know if the Apps plugin is available at this time, and we use this request as the check to determine this. Should we separate the concept of "the system is configured to use Apps via feature flag" and "the Apps plugin is enabled" from each other?

Merging master into this branch and resolving conflicts related to RHS bindings has brought this point up. If the frontend is reconnecting to the server, should we always try to fetch bindings, including RHS bindings? Because we don't know if the Apps plugin is available until we try to ping it. Maybe we should have a separate "ping" route that tells us if the plugin is running, so we can perform that request separately before fetching multiple sets of bindings.

File: packages/mattermost-redux/src/action_types/apps.ts [hunk 9]
```diff
@@ -7,7 +7,10 @@ import keyMirror from 'mattermost-redux/utils/key_mirror';
 // susceptible to breaking changes without pushing the major version of this package.
 export default keyMirror({
     RECEIVED_APP_BINDINGS: null,
+    FAILED_TO_FETCH_APP_BINDINGS: null,
     RECEIVED_APP_RHS_BINDINGS: null,
     RECEIVED_APP_COMMAND_FORM: null,
     RECEIVED_APP_RHS_COMMAND_FORM: null,
```
Comment (mickmister) [inline packages/mattermost-redux/src/action_types/apps.ts:13]: @larkox Do we need to add `FAILED_TO_FETCH_APP_RHS_BINDINGS` etc?

File: actions/websocket_actions.jsx [hunk 10]
```diff
@@ -1448,6 +1451,25 @@ function handleRefreshAppsBindings() {
...
 }
 
+export function handleAppsPluginEnabled() {
+    return (doDispatch, doGetState) => {
+        doDispatch({
+            type: AppsTypes.APPS_PLUGIN_ENABLED,
+        });
+
+        const state = doGetState();
+        doDispatch(fetchAppBindings(getCurrentUserId(state), getCurrentChannelId(state)));
```
Comment (larkox) [inline actions/websocket_actions.jsx:1461]: Probably not a big deal, but can we batch these dispatches?

File: packages/mattermost-redux/src/action_types/apps.ts (hunk 9 above)
Comment (larkox) [reply packages/mattermost-redux/src/action_types/apps.ts:13]: If I remember correctly, we are cleaning up before fetching the RHS data, so should not be needed. Nevertheless, I wonder if we better be safe than sorry, specially if we want to interpret a failure as "the plugin is not working anymore".

Also, we could keep the same action for both. I don't see material difference between failing to get the bindings on the RHS and failing to get the bindings on the main channel.

File: packages/mattermost-redux/src/reducers/entities/apps.ts (hunk 7 above)
Comment (larkox) [reply packages/mattermost-redux/src/reducers/entities/apps.ts:106]: More than a problem with the Apps plugin, what I foresee are problems either with the server (server is down) or with the client (client connection is lost).

We could tie fetching the bindings to the websocket reconnection (not sure if we already do), and therefore updating the apps enabled value on reconnection.

**Review by larkox — APPROVED** [review]

File: actions/websocket_actions.jsx (hunk 10 above)
Comment (mickmister) [reply actions/websocket_actions.jsx:1461]: @larkox I decided not to batch the actions because:

- it made testing that bindings were being fetched a little more complicated
- having separate actions dispatched makes it easier to debug in redux dev tools

File: packages/mattermost-redux/src/reducers/entities/apps.ts (hunk 7 above)
Comment (mickmister) [reply packages/mattermost-redux/src/reducers/entities/apps.ts:106]: @larkox Working on this task for mobile, I'm actually thinking now that we shouldn't invalidate the bindings for the "plugin disable" websocket event. If the user is looking at a thread on mobile, and the Apps plugin is toggled, the thread bindings would be cleared and wouldn't appear again right?

Comment (larkox) [reply packages/mattermost-redux/src/reducers/entities/apps.ts:106]: If we clear everything, the binding will dissappear, which is expected and desired, since we don't want the user to click on a binding that does not work. If the plugin gets enabled again, we could consider doing some logic to guess if we are in a thread, and bring the bindings, but that is not so critical (the end user does not need to have the bindings the instant the plugin gets online, and all users asking for the bindings at the same time could be really bad for the server).

File: actions/websocket_actions.jsx (hunk 10 above)
Comment (larkox) [reply actions/websocket_actions.jsx:1461]: Now that I think about this deeper, we shouldn't do this. This would imply that the moment we enable the plugin, all active users will ask the server for the plugin bindings, and that is costly. Since enabling or disabling the plugin is not something "critical" for the end user, we can wait until the user switch channels to get the bindings. What do you think?

File: packages/mattermost-redux/src/reducers/entities/apps.ts [hunk 11]
```diff
@@ -77,7 +91,28 @@ const rhs = combineReducers({
...
```
Comment (larkox) [inline packages/mattermost-redux/src/reducers/entities/apps.ts:106]: If the status is set to false, when do we expect to become true again? I understand one is on plugin enabled, but are we trying in any other place to see if the plugin is enabled?

Should we make a ping function for when the client gets "back online"? (Getting back from the background, recovering the internet connection...)

**Review by larkox — CHANGES_REQUESTED** [review]
One minor change, and another proposal. The change has to be done also for mobile, probably.

File: actions/websocket_actions.jsx (hunk 10 above)
Comment (mickmister) [reply actions/websocket_actions.jsx:1461]: @larkox Thinking about it more with your input, we can simply mark the plugin status as enabled in the reducer when the plugin is enabled, and let the user carry on without immediately fetching new bindings.

This is a similar situation to how we operate with webapp plugins. When a given plugin is enabled, every user fetches that plugin's webapp bundle simultaneously. A big difference here is that this operation is just fetching javascript/css files. The bindings call has much more dynamic computations involved.

- When an App is installed, should we not tell all users to refresh bindings?
- Should ee should allow the frontend to request bindings for a specific App so as to not storm all of the Apps at anytime we have an "all users" refresh situation
- 2/5 Maybe we should _always_ avoid making any situation where all users refresh bindings at the same time. So no App has to be able to handle all of those requests at the same time.

Something else that comes to mind: If an App has no channel-specific bindings, we can tell the Apps plugin to not fetch that App's bindings on channel switches. The frontend can create "soft" bindings requests for channel switches, and the Apps plugin would only fetch bindings from the Apps that are registered to have channel-specific bindings. The frontend also creates "hard" bindings requests, when it knows that it needs to do a full refresh of bindings. Then the Apps plugin would fetch from all Apps.

To accomplish this, we would need to include information in the call response about all Apps that are enabled, and which bindings can be cached that the frontend already has access to. The frontend also needs to know when to invalidate these bindings. We can send a hash of the cached response in the bindings request to the plugin, and have the plugin check if that hash is the last response computed for that user. The App or Apps plugin would be responsible for creating and storing this hash. The Apps plugin will read the stored hash and compare with what the frontend is providing in its request. What do you think?

cc @hanzei

File: packages/mattermost-redux/src/reducers/entities/apps.ts (hunk 11 above)
Comment (mickmister) [reply packages/mattermost-redux/src/reducers/entities/apps.ts:106]: @larkox there is a call to fetch bindings when the client reconnects after being offline (in this PR and the mobile PR). This doubles as the check to see if the plugin is enabled.

Maybe we should avoid the call to fetch bindings on reconnect and instead just ping the Apps plugin to see if it's enabled. This would avoid the situation where all users request bindings on reconnect simultaneously, e.g. when the server is restarted. cc @hanzei

File: actions/websocket_actions.jsx (hunk 10 above)
Comment (larkox) [reply actions/websocket_actions.jsx:1461]: I agree on not asking for bindings from all users at the same time.

Regarding caching, we've talked a lot about that in the past, and everything has its ups and downs. Mainly, we need to trust the app developer to be conscious about what of their bindings are cacheable, and give them a way to invalidate the caches, which is putting a lot more of effort on the app developer side for an implementation detail (cache).

File: packages/mattermost-redux/src/reducers/entities/apps.ts (hunk 11 above)
Comment (larkox) [reply packages/mattermost-redux/src/reducers/entities/apps.ts:106]: Regarding the second part of your comment, I sincerely don't know what direction to take. I feel that just pinging and setting the apps enabled value should be enough on reconnect (let them fetch the bindings on their next channel switch), but not sure.

File: actions/websocket_actions.jsx (hunk 10 above)
Comment (mickmister) [reply actions/websocket_actions.jsx:1461]: @levb Do you have any thoughts on the caching strategy discussed in the last two paragraphs of my comment above?

File: packages/mattermost-redux/src/reducers/entities/apps.ts (hunk 11 above)
Comment (mickmister) [reply packages/mattermost-redux/src/reducers/entities/apps.ts:106]: > Should we make a ping function for when the client gets "back online"?

@larkox I've added this function. Now the frontend reconnect handler just calls this `ping` route instead of fetching bindings. If the ping succeeds, bindings will be fetched at next channel switch.

**Review by larkox — APPROVED** [review]

File: actions/websocket_actions.jsx (hunk 10 above)
Comment (hanzei) [reply actions/websocket_actions.jsx]: Leaving the caching discussion aside, I'm wondering how much of an issue it is, when all clients as fetch bindings "at the same time".

First of call, I would like to point out that "at thee same time" can span over multiple seconds. Clients needs to receive the websocket event (variant delay), process and respond (variant delay) to it. Depending on the network this can take a couple of seconds.
For requests to the mm server and the apps plugins, the load doesn't seem like an issue. It's the same as for enabling a plugin. For the app itself processing all of these requests might be an issue, but the are naturally distributed for a few seconds. At least serverless infrastructure like AWS can handle this. HTTP apps might be an issue, but it's a deployment type that requires a load balancer anyway if run in production.

Comment (mickmister) [reply actions/websocket_actions.jsx]: @hanzei Can we discuss this in our meeting this morning?

**Comment by hanzei** [comment]
@mickmister What is the state of this PR? Is it ready for QA review?

File: packages/mattermost-redux/src/reducers/entities/apps.ts (hunk 11 above)
Comment (mickmister) [reply packages/mattermost-redux/src/reducers/entities/apps.ts:106]: We decided in a meeting that we would fetch bindings during reconnect, and not use a ping route to check for presence.

**Review by levb — APPROVED** [review]

**Comment by DHaussermann** [comment]
@mickmister I have confirmed with testing and upgrading that the behavior in the screenshot above seems to be a regression introduced by this PR. 

Feature Flag `true` AND plugin disabled will now not show the plugin as being available to enabled (without webapp refresh)

**Review by DHaussermann — APPROVED** [review]
Tested and passed. 
Retested the app when the feature Flag is true and the plugin is disabled as well as enabled.
When the plugin is disabled: 
- Re-tested that binding calls do not get sent when the app is enabled (Channel switch, Channel header, Post menu and command parser) 
- Confirmed issue above with plugin settings is resolved
- Bot and oAuth app data displays normally for data created by a plugin app 

LGTM!

**Comment by DHaussermann** [comment]
@amyblais Please note that this PR is required for the March 30th Cloud and v6.6 releases

**Comment by amyblais** [comment]
> @amyblais Please note that this PR is required for the March 30th Cloud and v6.6 releases

@DHaussermann @mickmister When will it be merged? This should be merged & cherry-picked to ``cloud`` by EOD today if we want it in March 30th Cloud and v6.6 releases, so SRE can update cloud test servers by tomorrow morning.

**Comment by DHaussermann** [comment]
@mickmister apologies, `Dev Review` label was re-added in error
